datalad_next/tree.py

```diff
@@ -0,0 +1,1130 @@
+            Path(path),
+            max_depth=depth,
+            exclude_node_func=build_excluded_node_func(
+                include_hidden=include_hidden, include_files=include_files),
+            **dataset_tree_args
+        )
+
+        for node in tree.generate_nodes():
+            # yield one node at a time to improve UX / perceived speed
+            res_dict = {
+                "action": "tree",
+                "path": str(node.path),
+                "type": node.TYPE,
+                "depth": node.depth,
+                "exhausted_levels": list(tree.exhausted_levels),
+                "count": {
+                    "datasets": tree.node_count["DatasetNode"],
+                    "directories": tree.node_count["DirectoryNode"],
+                    **({"files": tree.node_count["FileNode"]}
+                       if include_files else {})
+                },
+            }
+            if node.TYPE == "dataset":
+                res_dict.update({
+                    "dataset_depth": node.ds_depth,
+                    "dataset_abs_depth": node.ds_absolute_depth,
+                    "dataset_is_installed": node.is_installed
+                })
+
+            if node.is_symlink():
+                # TODO: should we inform if the symlink is recursive (as per
+                #  `tree.is_recursive_symlink()`) although not broken? The
+                #  UNIX 'tree' command shows the message '[recursive,
+                #  not followed]' next to the path. Not sure if this is
+                #  interesting at all or more confusing.
+                res_dict["symlink_target"] = node.get_symlink_target()
+                res_dict["is_broken_symlink"] = node.is_broken_symlink()
+
+            if node.exception is not None:
+                # mimic error message of unix 'tree' command for
+                # permission denied error, otherwise use exception short
+                # message
+                message = "error opening dir" \
+                    if node.exception.name == "PermissionError" \
+                    else node.exception.message
+
+                yield get_status_dict(
+                    status="error",
+                    message=message,
+                    exception=node.exception,
+                    **res_dict
+                )
+            else:
+                yield get_status_dict(
+                    status="ok",
+                    **res_dict
+                )
+
+    @staticmethod
+    def custom_result_renderer(res, **kwargs):
+        """
+        Each node is printed on one line. The string uses the format:
+        ``[<indentation>] [<branch_tip_symbol>] <path> [<ds_marker]``
+
+        Example line:
+        ``│   │   ├── path_dir_level3``
+        """
+        from datalad.support import ansi_colors
+
+        # get values from result record
+        node_type = res["type"]
+        node_path = res["path"]
+        depth = res["depth"]
+        exhausted_levels = res["exhausted_levels"]
```

Comment:
Here is an example result record from a deep tree, but for toplevel-ish directory:

```json
{
  "action": "tree",
  "count": {
    "datasets": 5,
    "directories": 9469
  },
  "dataset_abs_depth": 0,
  "dataset_depth": 0,
  "dataset_is_installed": true,
  "depth": 1,
  "exhausted_levels": [
    2,
    7,
    9,
    10,
    11,
    12,
    13,
    8,
    5,
    3,
    6,
    4
  ],
  "path": "/home/mih/dcm",
  "status": "ok",
  "type": "dataset"
}
```

If I get it right, the `exhausted_levels` are reported to understand for which branches of the tree to no longer draw `| ` prefix markers.

But in the case of a `depth=1` record, it should not be necessary to report any `exhausted_levels`, because they should all be exausted. In general, any report that is larger than the present `depth` is unnecessary.

As a "cosmetic" fix, we could do something like

```suggestion
        exhausted_levels = [l for l in res["exhausted_levels"] if l < depth]
```

but I am wondering whether this set should not be pruned in the tree node generator itself?